### ISSUE TITLE
O(n^2) complexity in kde-frameworks/extra-cmake-modules/setup-hook.sh

### ISSUE THREAD

**Issue by edolstra** [NixOS/nixpkgs#99308]
**Describe the bug**

The file `pkgs/development/libraries/kde-frameworks/extra-cmake-modules/setup-hook.sh` has this function:
```bash
ecmUnseenHostPath() {
    for pkg in "${ecmHostPathSeen[@]}"
    do
        if [ "${pkg:?}" == "$1" ]
        then
            return 1
        fi
    done

    ecmHostPathSeen+=("$1")
    return 0
}
```
Since this is called for every dependency, we get quadratic complexity. This may cause a non-trivial slowdown - building digikam I noticed it takes a few seconds just to get to `unpackPhase`, but there might be other causes. It also causes huge amounts of debug output when you add a `set -x` to `preHook` (about 600 MB for digikam).

We should use an associative array for this like stdenv does (see 8d76effc178747f0c8f456fe619c1b014736a6af).

**Notify maintainers**

@ttuegel

**Comment by ttuegel** [comment]
We also need to do this in the `qtbase` setup hook for paths seen by `qmake`.

**Comment by ttuegel** [comment]
Not stale.

**Comment by ttuegel** [comment]
Fixed in #154136.
